src/devices/sound/k051649.cpp

```diff
@@ -96,14 +96,50 @@ void k051649_device::device_reset()
 }
 
 
+//-------------------------------------------------
+//  device_post_load - device-specific post-load
+//-------------------------------------------------
+
+void k051649_device::device_post_load()
+{
+	device_clock_changed();
+}
+
+
+//-------------------------------------------------
+//  device_clock_changed - called if the clock
+//  changes
+//-------------------------------------------------
+
+void k051649_device::device_clock_changed()
+{
+	uint32_t old_rate = m_rate;
+	m_rate = clock()/16;
+	m_mclock = clock();
+	
+	if (old_rate < m_rate)
+	{
+		std::fill(m_mixer_buffer.begin(), m_mixer_buffer.end(), 0);
+		m_mixer_buffer.resize(2 * m_rate);
+	}
+	m_stream->set_sample_rate(m_rate);
+	if (old_rate > m_rate)
+	{
+		std::fill(m_mixer_buffer.begin(), m_mixer_buffer.end(), 0);
+		m_mixer_buffer.resize(2 * m_rate);
+	}
+}
+
+
 //-------------------------------------------------
 //  sound_stream_update - handle a stream update
 //-------------------------------------------------
 
 void k051649_device::sound_stream_update(sound_stream &stream, stream_sample_t **inputs, stream_sample_t **outputs, int samples)
 {
 	// zap the contents of the mixer buffer
-	memset(m_mixer_buffer.get(), 0, samples * sizeof(short));
+	std::fill(m_mixer_buffer.begin(), m_mixer_buffer.end(), 0);
+	short *mix;
```

Comment:
Why have you moved this variable here?  Logically the two `mix` variables are not related, and the code doesn't want the value from one channel to persist to the next channel.  It's not like it's an object that's expensive to construct.  But that aside, are either of the `mix` variables even needed?  Couldn't you just use `m_mixer_buffer[i]` in each case anyway, eliminating a piece of redundant state?